timemachine/cpp/src/potential.cu

```diff
@@ -1,16 +1,112 @@
-#include <iostream>
 #include <memory>
 
 #include "device_buffer.hpp"
 #include "fixed_point.hpp"
 #include "gpu_utils.cuh"
 #include "potential.hpp"
-#include "surreal.cuh"
+
+#include <chrono>
+#include <iostream>
 
 namespace timemachine {
 
 const int Potential::D = 3;
 
+void Potential::execute_batch_host(
+    const int coord_batches,     // Number of batches of coordinates
+    const int N,                 // Number of atoms
+    const int param_batches,     // Number of batches of parameters
+    const int P,                 // Number of parameters
+    const int lambda_batches,    // Number of lambda values
+    const double *h_x,           // [coord_batches, N, 3]
+    const double *h_p,           // [param_batches, P]
+    const double *h_box,         // [coord_batches, 3, 3]
+    const double *lambdas,       // [lambda_batches]
+    unsigned long long *h_du_dx, // [coord_batches, param_batches, lambda_batches, N, 3]
+    unsigned long long *h_du_dp, // [coord_batches, param_batches, lambda_batches, P]
+    unsigned long long *h_du_dl, // [coord_batches, param_batches, lambda_batches, N]
+    unsigned long long *h_u) {   // [coord_batches, param_batches, lambda_batches, N]
+    std::unique_ptr<DeviceBuffer<double>> d_p(nullptr);
+    if (P > 0) {
+        d_p.reset(new DeviceBuffer<double>(param_batches * P));
+        d_p->copy_from(h_p);
+    }
+
+    DeviceBuffer<double> d_box(coord_batches * D * D);
+    d_box.copy_from(h_box);
+
+    DeviceBuffer<double> d_x_buffer(coord_batches * N * D);
+    d_x_buffer.copy_from(h_x);
+
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dx_buffer(nullptr);
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dp_buffer(nullptr);
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_du_dl_buffer(nullptr);
+    std::unique_ptr<DeviceBuffer<unsigned long long>> d_u_buffer(nullptr);
+
+    const int K = coord_batches * param_batches * lambda_batches;
```

Comment:
Nit: possible confusion with later loop nest, where lower-case `k` indexes the innermost loop, rather than overall progress within the entire loop nest.

(Possibly better to use a different name for total number of iterations than upper case version of `i`, `j`, or `k`.)